plugins/experimental/carp/carp.cc

```diff
@@ -0,0 +1,713 @@
+        TSDebug(DEBUG_TAG_HOOK, "Carpable (%s) present but value not acceptable (%s)", CARPABLE_HEADER.c_str(), sCarpable.c_str());
+        TSHandleMLocRelease(bufp, hdr_loc, url_loc);
+        TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+        return 0;
+      } else {
+        TSDebug(DEBUG_TAG_HOOK, "Found Carpable header, routing...");
+      }
+    }
+  } else { // no whitelist so blacklist could be used
+    if (g_CarpConfigPool->getGlobalConfig()->isBlackListed(reqUrl.getHost())) { // if host black listed, do not carp
+      TSDebug(DEBUG_TAG_HOOK, "Host '%s' is blacklisted, not going through carp", reqUrl.getHost().c_str());
+      TSHandleMLocRelease(bufp, hdr_loc, url_loc);
+      TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+      return 0;
+    }
+  }
+  
+  TSDebug(DEBUG_TAG_HOOK, "URL to hash with=%s", sUrl.c_str());
+
+  // get nodeList and select node to forward to
+  std::vector<HashNode *> nodeList = g_CarpConfigPool->getGlobalHashAlgo()->getRemapProxyList(sUrl);
+  bool bAddHeaderResult=false;
+  bool bAddForwardedResult = false;
+  HashNode* node = NULL;
+  bool bIsOwner = false;
+
+  if (nodeList.size() == 0) { // no hosts available to forward to
+    TSDebug(DEBUG_TAG_HOOK, "no hosts available to forward to, will handle locally");
+    goto done;
+  } else {
+    node = nodeList[0];
+  }
+
+  bIsOwner = checkListForSelf(nodeList);
+  for (size_t k = 0; k < nodeList.size(); k++) {
+    TSDebug(DEBUG_TAG_HOOK, "nodeList host %d name is %s", static_cast<int>(k), nodeList[k]->name.c_str());
+  }
+
+  //handle forwarding
+  TSDebug(DEBUG_TAG_HOOK, "forwarding to '%s' (isSelf=%d)", node->name.c_str(), node->isSelf);
+  if (!node->isSelf) { // carp does not forward if we choose ourself
+    node->carpForward();
+    TSDebug(DEBUG_TAG_HOOK, "carp forwarded to %s.", node->name.c_str());
+    // insert carp loop prevention header
+    bAddHeaderResult = addHeader(bufp, hdr_loc, CARP_ROUTED_HEADER, string("1"));
+    if (!bAddHeaderResult) {
+      TSError("Carp, error inserting '%s' header", CARP_ROUTED_HEADER.c_str());
+    }
+    bAddForwardedResult = addHeader(bufp, hdr_loc, CARP_STATUS_HEADER, string(CARP_FORWARDED));
+    if (!bAddForwardedResult) {
+      TSError("Carp, error inserting '%s' header", CARP_STATUS_HEADER.c_str());
+    }
+
+    if (bIsPOSTRemap) { // if post remap, get remapped/OS Server Addr and add as header
+      const struct sockaddr* sa = TSHttpTxnServerAddrGet(txnp);
+      //        const struct sockaddr* sa = TSHttpTxnNextHopAddrGet(txnp);
+      if (sa) { // sanity check
+        struct sockaddr_storage ss;
+        memcpy(static_cast<void *> (&ss), sa, sizeof (sockaddr_storage));
+        if ((reinterpret_cast<sockaddr_in *> (&ss))->sin_port == 0) { // set port from client request URL
+          (reinterpret_cast<sockaddr_in *> (&ss))->sin_port = htons(reqUrl.getPort());
+        }
+
+        string sTemp;
+        getStringFromSockaddr(reinterpret_cast<const sockaddr *> (&ss), sTemp);
```

Comment:
This is so messed up I don't know where to start.

* As noted above, `sa` is certain to not be as long as a `sockaddr_storage` this prints out effectively random bytes to be shipped to the peer, presuming it doesn't just ASAN crash first.
* Since in other places it is clear only IPv4 is supported, why not use `sockaddr_in`? What's the benefit of using `soockaddr_storage`?
* Why bother with `ss` at all? AFAICT it's used only to set the port and since it's just printed to the string anyway, why not just print the port explicitly?
* Why not use `sTemp`? Or have the debug message use `val`. Having two different string conversions just for a debug output seems overly complex.
* In preference to either, why not use `inet_ntoa` or `inet_ntop`?